### ISSUE TITLE
Memory leakage due to tf.Variables

### ISSUE THREAD

**Issue by GlaserN** [q-optimize/c3#38]
Currently memory in optimization is strongly increasing with every iteration. Using a memory profiler this tracks down to redifining tf.Variables in the `propagation` of method in tensorflow:
https://github.com/q-optimize/c3/blob/ebb13e49f004551efb24b65ef1a746ebf7ba180c/c3/experiment.py#L333-L345
Here are memory profiling of 20 iterations using 
a) `dt = tf.variable...`
![memory_Variable](https://user-images.githubusercontent.com/20336812/106717010-66299880-65ff-11eb-8142-5cd92625d7f6.png)

and b) `dt =tf.constant`
![memory_constant](https://user-images.githubusercontent.com/20336812/106717026-6a55b600-65ff-11eb-9bba-0891f7647831.png)

It therefore seems that there is memory leakage due to the change to tf.Variables as introduced in  46be03f860ea080a3e14cd24dce45eda5d9e873f by @lazyoracle  what has been the main point of changing the definition. If there was no  specific reason, except for explicitely having to watch `constant` in the gradient tape I would suggest to revert tf.Variables to tf.constant.

**Comment by lazyoracle** [comment]
The original motivation for switching to `tf.Variable` was that it allowed the flexibility to directly access the whole suite of Gradient based optimizers defined in [`tf.keras.optimizer`](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer). I can go back and check if there is another way to continue that integration without fully rewriting parts of the tensorflow related codebase.

**Comment by fedroy** [comment]
We should integrate the `tf.Variable` with the Paramatermap and C3Obj, such that only the elements that need gradients are made into variables, the others are made `tf.constant`

**Comment by lazyoracle** [comment]
Possibly closed in #46
